### PR TITLE
BuildCheck replay mode - replaying events to logging service with registered BuildCheck logger

### LINKED ISSUE #9760: BuildCheck connector for binlog replaying
### Context

#9627

Today we have [`BuildCopConnectorLogger.cs`](https://github.com/dotnet/msbuild/blob/exp/build-analyzers/src/Analyzers/Infrastructure/BuildCopConnectorLogger.cs) as a connector between Logging infrastructure and BuildInfrastructure. It Implements the MSBuild `ILogger` interface and hance can be plugged in as logger, receive the `BuildEventArgs` data and push those to the [`BuildCopManager.cs`](https://github.com/dotnet/msbuild/blob/exp/build-analyzers/src/Analyzers/Infrastructure/BuildCopManager.cs).

The MSBuild is able to read the binlog and replay it to any attached loggers (via `MSBuild.exe <binlogpath> <other args - to attach loggers etc>`). 

### Goal

We aim to leverage both facts in 'BuildCop post-build mode' - MSBuild will replay `BuildEventArgs` from binlog and send them to BuildCop that will report it's findings.

### Subproblems

(Order is not strict)

 - [ ] Author an end-to-end unit test that would create binlog, then replay it in BuildCop mode and assert that report is outputed
 - [ ] Identify and fix or plan fix for any issues with analysis during replay mode
 - [ ] Make sure .editorconfig files (touched by BuildCop, or possibly all) in the source tree that is build with binlogger attached (`/bl`) are embedded into binlog file.
-----------------
 - [ ] In replay mode - use .editorconfig files from the binlog (if any).
 - [ ] The build might have happen on top of sources without .editorconfig files, while in replay mode we might want to tune the rules - so add an option to specify editorconfig file during replay (e.g. via command line switch)
 - [ ] (skip now) There will be Analyzers, that will need to hook directly into executing nodes (the way of distinguishing those is yet TBD) - those will need to be skipped during replay analysis mode, but some message/warning will need to be output to inform about this. In a future version we might as well reconstruct some of those 'execution-specific' data during replay mode (e.g. we can reconstruct `ProjectInstance`/`Project` from the data embedded in binlog)

### PR DESCRIPTION
**Pull request by surayya-MS**
Fixes #9760

Draft solution for BuildCheck Replay Mode.

If BuildCheck is enabled then create loggingService with attached BuildCheck logger and initialize it for replaySource. When events are replayed to loggingService, the BuildCheck logger issues new events that loggingService hanled accordingly. This is similar to what's happennning during build process.

This is probably not the approach we're going for.

### Notes
If wee need to fiter out events that we replay, it is easy to accomplish with this solution.

### CODE REVIEW

File: src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs [hunk 1]
```diff
@@ -597,7 +598,14 @@ private void WriteBaseFields(BuildEventArgs e, BuildEventArgsFieldFlags flags)
         {
             if ((flags & BuildEventArgsFieldFlags.Message) != 0)
             {
-                WriteDeduplicatedString(e.RawMessage);
+                if (e is BuildCheckResultWarning)
```
Comment (JanKrivanek) [inline src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs:601]: The other result types will have similar problem.
But let's try to think if we can solve this without special-casing - we'll probably need to to design the BuildCheck results events more properly

File: src/Build/BackEnd/BuildManager/BuildManager.cs [hunk 2]
```diff
@@ -2951,10 +2951,38 @@ private void OnProjectStarted(object sender, ProjectStartedEventArgs e)
...
+            ISet<string> warningsNotAsErrors,
+            ISet<string> warningsAsMessages)
+        {
+            _buildParameters = buildParameters;
+
+            _workQueue ??= new ActionBlock<Action>(action => ProcessWorkQueue(action));
+
+            lock (_syncLock)
+            {
+                AttachDebugger();
```
Comment (JanKrivanek) [inline src/Build/BackEnd/BuildManager/BuildManager.cs:2969]: nit: leftover

**Review by JanKrivanek — COMMENTED** [review]
I like the simplicity of this approach.
Though the need for filtering out previous reports might complitate this solution a bit.

As discussed - I'd preffer the build check to pluch as an event source proxy (basically taking IEventSource as input and producing IEventSource as output) - that way no components from the build would need to know about it (now BuildCheckManager and Logger (for filtering) would need to have knowledge of the build check replay scenario)

**Comment by surayya-MS** [comment]
Looks like there is a compatibility problem 
```
.dotnet\sdk\8.0.201\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error CP0007: (NETCORE_ENGINEERING_TELEMETRY=Build) Type 'Microsoft.Build.Logging.BinaryLogReplayEventSource' does not inherit from base type 'Microsoft.Build.Logging.EventArgsDispatcher' on ref/net472/Microsoft.Build.dll but it does on [Baseline] ref/net472/Microsoft.Build.dll
```
I will change the implementation slightly to make it compatible

File: src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs (hunk 1 above)
Comment (surayya-MS) [reply src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs:601]: done in this PR https://github.com/dotnet/msbuild/pull/10265

**Comment by surayya-MS** [comment]
Closing in favor of this PR https://github.com/dotnet/msbuild/pull/10224

**Pull request closed without merging**